[Console] Add getter for the original command "code" object

weitzman opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

I'm playing with Invokable Commands and am seeing a limitation which is bothersome. Attributes attached to the invokable class are lost by the time we have a  `Command` added to the Application. Consider an invokable class like below. The #[CLI\FieldLabels] and #[CLI\DefaultFields] Attributes are not retrievable from the Command. That is, there is no easy way to access the actual InvokableCommand object instead of the [wrapping Command](https://github.com/symfony/symfony/blob/a384c231a0051c0ac10e89c2a0516343f9fd3187/src/Symfony/Component/Console/Application.php#L553).

With this PR, Attributes become accessible via `$command->getCode()->getCallable()`

```php
namespace Drush\Commands\core;

#[AsCommand(
    name: 'twig:unused',
    description: 'Find potentially unused Twig templates.',
    aliases: ['twu'],
)]
#[CLI\FieldLabels(labels: ['template' => 'Template', 'compiled' => 'Compiled'])]
#[CLI\DefaultFields(fields: ['template', 'compiled'])]
final class TwigUnusedCommand
{
    public function __invoke(
        #[Argument(description: 'A comma delimited list of paths to recursively search')] string $searchpaths,
        InputInterface $input,
        OutputInterface $output
    ): int
    {
        $data = $this->doExecute($searchpaths);
        $this->writeFormattedOutput($input, $output, $data);
        return Command::SUCCESS;
    }
}
```

weitzman committed:
[Console] Add getter for the original command "code" object

weitzman committed:
[Console] Add getter for the original command "code" object

weitzman commented:
@derrabus - I added a hyperlink to the OP explaining whats meant by "wrapping command". Let me know if its still unclear.

If anyone can suggest an existing test to modify, I will do so.

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -356,6 +356,11 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
         }
     }
 
+    public function getCode(): ?InvokableCommand
```
The `InvokableCommand` class in *internal*, we have to change that in order to use it as return type hint in the public API.

chalasr replied:
I would keep it internal though

chalasr commented:
Jérôme has a good point about manipulating the command at compile time instead of runtime. When accessing the service you get the real instance .

GromNaN commented:
Thanks for the context. 

I see that Drush is built with `consolidation/annotated-command`. Looking at [its documentation](https://github.com/consolidation/annotated-command/blob/4.x/README.md#example-annotated-command-class) for the `FieldLabels` and `DefaultFields` attributes, it looks very close to what we built with invokable commands. I do think that the two attribute systems (Symfony and Consolidation) are mutually exclusive to configure the command.

weitzman commented:
We are working to remove the dependency on consolidation/annotated-command in favor native Symfony Console conventions. See https://github.com/drush-ops/drush/pull/6135. This Console PR is part of the effort.

weitzman commented:
I have added tests to the PR.

So, do we think this is solvable for us and other standalone Console users? I hope so.

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -356,6 +356,14 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
+    /**
+     * Get the InvokableCommand instance if it exists.
+     */
+    public function getCode()
+    {
+        return $this->code;
+    }
```
We can skip the internal `InvokableCommand` and return the callable that has been set.

```suggestion
    public function getCode(): ?callable
    {
        return $this->code?->getCallable();
    }
```

GromNaN commented on src/Symfony/Component/Console/Tests/Command/InvokableCommandTest.php:
```diff
@@ -292,6 +294,16 @@ public function __invoke()
+        $command = new Command(null, new InvokableTestCommand());
+
+        // Ensure that the invokable class can be retrieved from the Command.
+        $this->assertInstanceOf(InvokableCommand::class, $command->getCode());
+        $this->assertInstanceOf(InvokableTestCommand::class, $command->getCode()->getCallable());
```
```suggestion
        $this->assertInstanceOf(InvokableTestCommand::class, $command->getCode());
```

GromNaN reviewed (commented):
Here is a solution so keep the `InvokableCommand` internal.

Do you want to read the Consolidation attributes for backward compatibility? I still don't get how you will proceed.

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -356,6 +356,14 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
     }
 
+    /**
+     * Get the InvokableCommand instance if it exists.
```
The comment needs to be updated.

GromNaN commented on src/Symfony/Component/Console/Command/InvokableCommand.php:
```diff
@@ -34,13 +34,15 @@ class InvokableCommand implements SignalableCommandInterface
     public function __construct(
         private readonly Command $command,
-        callable $code,
+        callable $callable,
```
CPP can be used here.
```suggestion
        private callable $callable,
```

weitzman replied:
Surprisingly, PHP does not allow `callable` type hint on a property. See https://wiki.php.net/rfc/typed_properties_v2#supported_types and https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.base.function. Given this limitation, I think we keep the parameter type hint to callable and not use CPP. Sound OK?

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -356,6 +356,14 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
     }
 
+    /**
+     * Get the callable from the InvokableCommand if available.
```
`InvokableCommand` is internal, it should not be mentioned.


```suggestion
     * Get the code that is executed by the command.
```

stof commented on src/Symfony/Component/Console/Command/InvokableCommand.php:
```diff
@@ -34,13 +34,15 @@ class InvokableCommand implements SignalableCommandInterface
 
     public function __construct(
         private readonly Command $command,
-        callable $code,
```
I would not rename the argument here. Naming it `code` is clearer about what the callable is about.

weitzman replied:
I pushed some tweaks variable names. This one is now unchanged. I'm open to alternatives.

weitzman commented:
I think I've addressed all feedback. The one test failure is unrelated to this PR.

GromNaN commented on src/Symfony/Component/Console/Tests/Command/InvokableCommandTest.php:
```diff
@@ -292,6 +293,15 @@ public function __invoke()
+        $command = new Command(null, new InvokableTestCommand());
+
+        // Ensure that the invokable class can be retrieved from the Command.
+        $this->assertInstanceOf(InvokableTestCommand::class, $command->getCode());
```
You can assert it's the same object.

GromNaN reviewed (changes requested):
Please add an entry in the changelog of the console component.

weitzman commented:
Changelog entry added.

GromNaN commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 7.4
 ---
 
+ * A new $command->getCode() method enables getting the invokable object.
```
We use this syntax:

```suggestion
 * A new `Command::getCode()` method enables getting the invokable object.
```

antonkomarev commented:
Getter name `getCode` looks like status code. Those were my initial thoughts when I read the PR title.

GromNaN commented:
> Getter name `getCode` looks like status code. Those were my initial thoughts when I read the PR title.

This name is symmetrical to the existing `setCode`. If it was returning the status code, we would name it `getStatusCode`.

weitzman commented:
Anything else needed here before merge? Thanks all.

GromNaN commented:
@symfony/mergers Could you review this PR please?

fabpot commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -356,6 +356,14 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
     }
 
+    /**
+     * Get the code that is executed by the command.
```
```suggestion
     * Gets the code that is executed by the command.
     * 
     * @return ?callable null if the code has not been set with setCode()
```

fabpot commented on src/Symfony/Component/Console/Command/InvokableCommand.php:
```diff
@@ -30,26 +30,28 @@
 
     public function __construct(
         private readonly Command $command,
         callable $code,
```
```suggestion
        private callable $code,
```

weitzman replied:
When we make this change, PHP gives an error: "Property cannot have type 'callable'". See https://stackoverflow.com/questions/57935734/is-type-callable-supported-with-typed-properties.

GromNaN replied:
The `callable` pseudo-type cannot be used as property type. That's why CPP isn't used here.

https://github.com/symfony/symfony/pull/61078#discussion_r2197469426

fabpot commented on src/Symfony/Component/Console/Command/InvokableCommand.php:
```diff
@@ -30,26 +30,28 @@
     private readonly ?SignalableCommandInterface $signalableCommand;
     private readonly \ReflectionFunction $reflection;
     private bool $triggerDeprecations = false;
+    private $code;
```
```suggestion
```

weitzman replied:
Alas I dont think this will work. See my comment below in __construct() about callable in typed property.

fabpot commented on src/Symfony/Component/Console/Command/InvokableCommand.php:
```diff
@@ -30,26 +30,28 @@
         callable $code,
     ) {
-        $this->code = $this->getClosure($code);
+        $this->code = $code;
```
```suggestion
```

weitzman replied:
Alas I dont think this will work. See my comment above about `callable` in typed property

fabpot commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 7.4
 ---
 
+ * A new `Command::getCode()` method enables getting the invokable object.
```
```suggestion
 * Add `Command::getCode()` to get the code set via `setCode()`.
```

alexandre-daubois commented on src/Symfony/Component/Console/Tests/Command/InvokableCommandTest.php:
```diff
@@ -293,6 +294,16 @@ public function __invoke()
 
+    public function testGetCode()
+    {
+        // Create a command from an invokable class.
```
I don't feel like this comment adds value

alexandre-daubois commented on src/Symfony/Component/Console/Tests/Command/InvokableCommandTest.php:
```diff
@@ -293,6 +294,16 @@ public function __invoke()
+        $invokableTestCommand = new InvokableTestCommand();
+        $command = new Command(null, $invokableTestCommand);
+
+        // Ensure that the invokable class can be retrieved from the Command.
```
Ditto

GromNaN committed:
feature #61078 [Console] Add getter for the original command "code" object (weitzman)

This PR was squashed before being merged into the 7.4 branch.

Discussion
----------

[Console] Add getter for the original command "code" object

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

I'm playing with Invokable Commands and am seeing a limitation which is bothersome. Attributes attached to the invokable class are lost by the time we have a  `Command` added to the Application. Consider an invokable class like below. The #[CLI\FieldLabels] and #[CLI\DefaultFields] Attributes are not retrievable from the Command. That is, there is no easy way to access the actual InvokableCommand object instead of the [wrapping Command](https://github.com/symfony/symfony/blob/a384c231a0051c0ac10e89c2a0516343f9fd3187/src/Symfony/Component/Console/Application.php#L553).

With this PR, Attributes become accessible via `$command->getCode()->getCallable()`

```php
namespace Drush\Commands\core;

#[AsCommand(
    name: 'twig:unused',
    description: 'Find potentially unused Twig templates.',
    aliases: ['twu'],
)]
#[CLI\FieldLabels(labels: ['template' => 'Template', 'compiled' => 'Compiled'])]
#[CLI\DefaultFields(fields: ['template', 'compiled'])]
final class TwigUnusedCommand
{
    public function __invoke(
        #[Argument(description: 'A comma delimited list of paths to recursively search')] string $searchpaths,
        InputInterface $input,
        OutputInterface $output
    ): int
    {
        $data = $this->doExecute($searchpaths);
        $this->writeFormattedOutput($input, $output, $data);
        return Command::SUCCESS;
    }
}
```

Commits
-------

b57d27544e6 [Console] Add getter for the original command "code" object
